src/main/scala/is/hail/utils/richUtils/RichRDD.scala

```diff
@@ -36,34 +36,44 @@ class RichRDD[T](val r: RDD[T]) extends AnyVal {
     hConf.delete(filename, recursive = true) // overwriting by default
 
     val parallelOutputPath =
-      if (parallelWrite) {
-        filename
-      } else
+      if (exportType == ExportType.CONCATENATED)
         hConf.getTemporaryFile(tmpDir)
+      else
+        filename
 
     val rWithHeader = header.map { h =>
-      if (r.getNumPartitions == 0)
+      if (r.getNumPartitions == 0 && exportType != ExportType.PARALLEL_SEPARATE_HEADER)
         r.sparkContext.parallelize(List(h), numSlices = 1)
-      else if (parallelWrite)
-        r.mapPartitions { it => Iterator(h) ++ it }
-      else
-        r.mapPartitionsWithIndex { case (i, it) =>
-          if (i == 0)
-            Iterator(h) ++ it
-          else
-            it
+      else {
+        exportType match {
+          case ExportType.CONCATENATED =>
+            r.mapPartitionsWithIndex { case (i, it) =>
+              if (i == 0)
+                Iterator(h) ++ it
+              else
+                it
+            }
+          case ExportType.PARALLEL_SEPARATE_HEADER =>
+            r
+          case ExportType.PARALLEL_HEADER_IN_SHARD =>
+            r.mapPartitions { it => Iterator(h) ++ it }
+          case _ => fatal(s"Unknown export type: $exportType")
         }
+      }
     }.getOrElse(r)
 
     codec match {
       case Some(x) => rWithHeader.saveAsTextFile(parallelOutputPath, x.getClass)
       case None => rWithHeader.saveAsTextFile(parallelOutputPath)
     }
 
+    if (exportType == ExportType.PARALLEL_SEPARATE_HEADER)
+      hConf.writeTextFile(parallelOutputPath + "/header")(out => out.write(header.getOrElse("")))
```

Comment:
How do I incorporate the codec here?

Reply:
I think the header will need to have the right extension. You can get the codec with similar code to `RichHadoopConfiguration.stripCodec`:
```
  def stripCodec(s: String): String = {
    val path = new org.apache.hadoop.fs.Path(s)

    Option(new CompressionCodecFactory(hConf)
      .getCodec(path))
      .map { codec =>
        val ext = codec.getDefaultExtension
        assert(s.endsWith(ext))
        s.dropRight(ext.length)
      }.getOrElse(s)
  }
```

Then append it to the header.